Allow storage options to be passed

bolkedebruin opened the pull request:
This allows storage options to be passed on to the underlying implementation. It does require a release of the providers as the signature of `get_fs` has changed.

@kaxil @uranusjr @Taragolis 


<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

bolkedebruin committed:
Allow storage options to be passed

Taragolis commented on airflow/providers/amazon/aws/fs/s3.py:
```diff
@@ -45,7 +45,7 @@ class SignError(Exception):
 
 
-def get_fs(conn_id: str | None) -> AbstractFileSystem:
+def get_fs(conn_id: str | None, storage_options: dict[str, str] | None) -> AbstractFileSystem:
```
```suggestion
def get_fs(conn_id: str | None, storage_options: dict[str, str] | None = None) -> AbstractFileSystem:
```

Taragolis commented on airflow/providers/google/cloud/fs/gcs.py:
```diff
@@ -39,7 +39,7 @@
 
 
-def get_fs(conn_id: str | None) -> AbstractFileSystem:
+def get_fs(conn_id: str | None, storage_options: dict[str, str]) -> AbstractFileSystem:
```
```suggestion
def get_fs(conn_id: str | None, storage_options: dict[str, str] = None) -> AbstractFileSystem:
```

bolkedebruin replied:
I don't think mypy will pass on this :-)

Taragolis commented on airflow/providers/microsoft/azure/fs/adls.py:
```diff
@@ -27,7 +27,7 @@
 
 
-def get_fs(conn_id: str | None) -> AbstractFileSystem:
+def get_fs(conn_id: str | None, storage_options: dict[str, Any] | None) -> AbstractFileSystem:
```
```suggestion
def get_fs(conn_id: str | None, storage_options: dict[str, Any] | None = None) -> AbstractFileSystem:
```

bolkedebruin committed:
Update airflow/providers/amazon/aws/fs/s3.py

Co-authored-by: Andrey Anshin <Andrey.Anshin@taragol.is>

bolkedebruin committed:
Update airflow/providers/microsoft/azure/fs/adls.py

Co-authored-by: Andrey Anshin <Andrey.Anshin@taragol.is>

bolkedebruin committed:
Add none as default

eladkal reviewed (approved):
LGTM

bolkedebruin committed:
Temp

uranusjr committed:
Pass conn ID to ObjectStoragePath via URI (#35913)

This enables an alternative ObjectStoragePath init syntax, using the
auth section in the URI to supply conn ID instead of a separate keyword
argument. The explicit keyword argument is honored if supplied.

bolkedebruin committed:
Reinstate changes

bolkedebruin committed:
Reapply patches

bolkedebruin committed:
revert encrypt in serde for this pr

uranusjr commented on airflow/io/__init__.py:
```diff
@@ -32,21 +32,25 @@
 if TYPE_CHECKING:
     from fsspec import AbstractFileSystem
 
+    from airflow.io.typedef import Properties
```
Why does this need a dedicated type alias? Do we plan to make it stricter with say TypedDict? It feels to me it’s better to use `dict[str, str]` instead; slightly repetitive but cognatively easier for readers.

bolkedebruin replied:
I like the 'signal' of Properties that it automatically says I can only pass [str, str] instead. To me it feels cognitive easier to see Properties than dict[str, str] which has no semantic meaning and less error prone. But I am not really tied up on it. 

And yes maybe we could make it into a TypedDict

what do you say?

uranusjr commented on airflow/io/__init__.py:
```diff
@@ -32,21 +32,25 @@
-def _file(_: str | None) -> LocalFileSystem:
-    return LocalFileSystem()
+def _file(_: str | None, storage_options: Properties) -> LocalFileSystem:
+    options = storage_options or {}
```
This seems unnecessary since properties are always a dict

uranusjr commented on airflow/io/store/__init__.py:
```diff
@@ -134,6 +146,6 @@ def attach(
         if store := _STORE_CACHE.get(alias):
             return store
 
-    _STORE_CACHE[alias] = store = ObjectStore(protocol=protocol, conn_id=conn_id, fs=fs)
+    _STORE_CACHE[alias] = store = ObjectStore(protocol=protocol, conn_id=conn_id, fs=fs, **kwargs)
```
It makes me uncomfortable that storage options are only effective in the first call to `attach`; in later calls they will be silently ignored, even when options in the existing storage object does not match. It would be more explicit (thus less prone to user errors) if storage creation and retrieval should be splitted into two functions instead.

bolkedebruin replied:
Yeah, you have a point. Do you have a suggestion?

bolkedebruin committed:
Remove encryption

bolkedebruin committed:
Fix typing
